internal/controller/servicetemplate_controller.go

```diff
@@ -0,0 +1,580 @@
+		},
+	}
+	key := client.ObjectKeyFromObject(gitRepository)
+	if err := r.Get(ctx, key, gitRepository); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get GitRepository %s: %w", key, err)
+		}
+	}
+
+	op, err := controllerutil.CreateOrUpdate(ctx, r.Client, gitRepository, func() error {
+		gitRepository.SetLabels(map[string]string{
+			kcm.KCMManagedLabelKey: kcm.KCMManagedLabelValue,
+		})
+		gitRepository.Spec = sourcev1.GitRepositorySpec{
+			Provider:          ref.Provider,
+			Interval:          ref.Interval,
+			Timeout:           ref.Timeout,
+			Ignore:            ref.Ignore,
+			Suspend:           ref.Suspend,
+			URL:               ref.Git.URL,
+			Reference:         ref.Git.Reference,
+			Verification:      ref.Git.Verification,
+			RecurseSubmodules: ref.Git.RecurseSubmodules,
+			Include:           ref.Git.Include,
+		}
+		if template.Spec.Authorization != nil {
+			gitRepository.Spec.SecretRef = template.Spec.Authorization.SecretRef
+			gitRepository.Spec.ProxySecretRef = template.Spec.Authorization.ProxySecretRef
+		}
+		return controllerutil.SetControllerReference(template, gitRepository, r.Client.Scheme())
+	})
+	if err != nil {
+		return fmt.Errorf("failed to reconcile GitRepository object: %w", err)
+	}
+	if op == controllerutil.OperationResultCreated || op == controllerutil.OperationResultUpdated {
+		ctrl.LoggerFrom(ctx).Info("Successfully mutated GitRepository", "GitRepository", client.ObjectKeyFromObject(gitRepository), "operation_result", op)
+	}
+	if op == controllerutil.OperationResultNone {
+		conditions := make([]metav1.Condition, len(gitRepository.Status.Conditions))
+		copy(conditions, gitRepository.Status.Conditions)
+		template.Status.SourceStatus = &kcm.SourceStatus{
+			Kind:               "GitRepository",
+			Name:               gitRepository.Name,
+			Namespace:          gitRepository.Namespace,
+			Artifact:           gitRepository.Status.Artifact,
+			ObservedGeneration: gitRepository.Generation,
+			Conditions:         conditions,
+		}
+		template.Status.Valid = slices.ContainsFunc(gitRepository.Status.Conditions, func(c metav1.Condition) bool {
+			return c.Type == conditionReady && c.Status == metav1.ConditionTrue
+		})
+	}
+	return nil
+}
+
+func (r *ServiceTemplateReconciler) reconcileBucket(
+	ctx context.Context,
+	template *kcm.ServiceTemplate,
+	ref *kcm.RemoteSourceSpec,
+) error {
+	// we'll create a new Bucket object for every ServiceTemplate. This is needed because we cannot reuse the
+	// same Bucket objects for multiple ServiceTemplates due to possible difference between bucket parameters.
+	// Since there would be no labels to identify created Bucket objects, they won't be reconciled by the flux
+	// controller. The OwnerReference will be set to the ServiceTemplate object, so that the Bucket object will be
+	// deleted when the ServiceTemplate is deleted.
+	bucketName, err := GenerateSourceName(client.ObjectKeyFromObject(template))
```

Comment:
Generating random name works, but just wondering, if there is a 1-to-1 relationship between a servicetemplate and bucket, will it be more user-friendly to generate the name as `<servicetemplate-name>-bucket` instead?